vulpes/upgrades/gamestate.cpp

```diff
@@ -0,0 +1,423 @@
+    //game_state_globals_file_handle = *reinterpret_cast<HANDLE**>(write_to_file_patch_addr+2);
+
+    // Patch the original table allocation function to replace it with ours.
+    patch_gamestate_new_replacement.build(sig_game_state_data_new());
+    patch_gamestate_new_replacement.apply();
+
+    // Make sure to also load our file when the game loads from the main checkpoint file.
+    patch_gamestate_read_from_main_file_hook.build(sig_game_state_read_from_main_file_hook());
+    patch_gamestate_read_from_main_file_hook.apply();
+
+    // Also load our file from the player specific checkpoint folder when the game does so.
+    patch_gamestate_read_from_profile_file_hook.build(sig_game_state_read_from_profile_file_hook());
+    patch_gamestate_read_from_profile_file_hook.apply();
+
+    // Hook into a small piece of code in the checkpoint file writing code
+    // so we can execute our file writing code.
+    patch_gamestate_write_to_files_hook.build(sig_game_state_write_to_file_patch());
+    patch_gamestate_write_to_files_hook.apply();
+
+    // Replace parts of the code that handles the memory copy to the checkpoint
+    // buffer with our own code.
+    patch_copy_to_checkpoint_state_hook.build(sig_game_state_copy_to_checkpoint_buffer());
+    patch_copy_to_checkpoint_state_hook.apply();
+
+    // Extend the checkpoint file copying function to also copy our checkpoint
+    // file.
+    patch_copy_checkpoint_file_hook.build(sig_copy_checkpoint_file_hook());
+    patch_copy_checkpoint_file_hook.apply();
+    gamestate_copy_checkpoint_file_continue_ptr = patch_copy_checkpoint_file_hook.return_address();
+
+    // This is the location the game stores its gamestate in.
+    uintptr_t mem_map_loc = *sig_physical_memory_map_location();
+
+    // Allocate and null the memory for our upgrades.
+    gamestate_extension_buffer = VirtualAlloc(
+        // Allocating on a static address relieves us from the duty of needing
+        // to patch loaded stuff to fit in the current memory model.
+        // We basically allocate our chunk of gamestate before vanilla gamestate.
+        reinterpret_cast<void*>(mem_map_loc-ALLOCATED_UPGRADE_MEMORY),
+        ALLOCATED_UPGRADE_MEMORY,
+        MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+    gamestate_extension_checkpoint_buffer = VirtualAlloc(
+        NULL,
+        ALLOCATED_UPGRADE_MEMORY,
+        MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+
+    assert(gamestate_extension_buffer);
+    assert(gamestate_extension_checkpoint_buffer);
+
+    memset(gamestate_extension_buffer, 0, ALLOCATED_UPGRADE_MEMORY);
+}
+
+void revert_gamestate_upgrades() {
+    if (!initialized) return;
+
+    patch_gamestate_new_replacement.revert();
+    patch_gamestate_write_to_files_hook.revert();
+    patch_copy_to_checkpoint_state_hook.revert();
+    patch_gamestate_read_from_main_file_hook.revert();
+    patch_gamestate_read_from_profile_file_hook.revert();
+    patch_copy_to_checkpoint_state_hook.revert();
```

Comment:
You don't revert these in the same or reverse order you apply them in. Is that ok?

Reply:
It's okay because they're not overlapped.